devito/petsc/iet/passes.py

```diff
@@ -0,0 +1,366 @@
+                                Vec, KSP, PC, SNES, PetscErrorCode, PETScArray)
+from devito.symbolics import Byref, Macro, FieldFromPointer
+from devito.petsc.iet.nodes import MatVecAction, LinearSolverExpression
+from devito.petsc.utils import linear_solver_mapper
+import cgen as c
+
+
+@iet_pass
+def lower_petsc(iet, **kwargs):
+
+    # Check if PETScSolve was used.
+    petsc_nodes = FindNodes(MatVecAction).visit(iet)
+
+    if not petsc_nodes:
+        return iet, {}
+
+    else:
+        # Collect all petsc solution fields
+        unique_targets = list(set([i.expr.rhs.target for i in petsc_nodes]))
+
+        # Initalize PETSc
+        init = init_petsc(**kwargs)
+
+        # Create context data struct
+        struct = build_struct(iet)
+
+        objs = build_core_objects(unique_targets[-1], **kwargs)
+
+        # Create core PETSc calls (not specific to each PETScSolve)
+        core = core_petsc(unique_targets[-1], struct, objs, **kwargs)
+
+        matvec_mapper = MapNodes(Iteration, MatVecAction, 'groupby').visit(iet)
+
+        main_mapper = {}
+
+        setup = []
+        efuncs = []
+
+        for target in unique_targets:
+
+            solver_objs = build_solver_objs(target)
+
+            matvec_body_list = List()
+
+            solver_setup = False
+
+            for iter, (matvec,) in matvec_mapper.items():
+
+                # Skip the MatVecAction if it is not associated with the target
+                # There will most likely be more than one MatVecAction
+                # associated with the target e.g interior matvec + BC matvecs
+                if matvec.expr.rhs.target != target:
+                    continue
+
+                # Only need to generate solver setup once per target
+                if not solver_setup:
+                    solver = generate_solver_calls(solver_objs, objs, matvec, target)
+                    setup.extend(solver)
+                    solver_setup = True
+
+                # Make the body of the matrix-vector callback for this target
+                matvec_body = matvec_body_list._rebuild(body=[
+                    matvec_body_list.body, iter[0]])
+                matvec_body_list = matvec_body_list._rebuild(body=matvec_body)
+
+                # Remove the iteration loop from the main kernel encapsulating
+                # the matvec equations since they are moved to the callback
+                main_mapper.update({iter[0]: None})
+
+            # Create the matvec callback and operation for each target
+            matvec_callback, matvec_op = create_matvec_callback(
+                target, matvec_body_list, solver_objs, objs,
+                struct)
+
+            setup.append(matvec_op)
+            setup.append(BlankLine)
+            efuncs.append(matvec_callback)
+
+        # Remove the LinSolveExpr from iet and efuncs that were used to carry
+        # metadata e.g solver_parameters
+        main_mapper.update(rebuild_expr_mapper(iet))
```

Comment:
for mappers fed to Transformer, or any *replace routines, we often use the name "subs", which frees up the name "mapper" for another object